casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -274,110 +279,143 @@ class Running[F[_]: Sync: BlockStore: CommUtil: TransportLayer: ConnectionsCell:
+    * bad and useless blocks. So the check is if block is well formed and of interest for this particular node.
+    * Should not catch any slashable offences - such blocks should be persisted and processed by Casper.
+    * @param b BlockMessage
+    * @param peer peer from which BlockMessage came from
+    * @return if block is suitable for this node
     */
-  private def casperAdd(peer: PeerNode)(b: BlockMessage): F[ValidBlockProcessing] = {
-    import cats.instances.option._
-
-    // Save block in block store immediately
-    BlockStore[F].put(b) >>
-      validatorId.traverse_ { id =>
-        val self = ByteString.copyFrom(id.publicKey.bytes)
-        F.whenA(b.sender == self)(
-          Log[F].warn(
-            s"There is another node $peer proposing using the same private key as you. Or did you restart your node?"
-          )
-        )
-      } >> {
-      for {
-        maybeReq <- RequestedBlocks.get(b.blockHash)
-        _ <- maybeReq match {
-              // There might be blocks that are not maintained by RequestedBlocks, e.g. genesis ceremony messages
-              case None =>
-                Log[F].info(
-                  s"Block ${PrettyPrinter.buildString(b.blockHash)} is not present in RequestedBlocks."
-                )
-              case Some(requested) => {
-                // Make Casper loop aware that the block has been received
-                RequestedBlocks.put(b.blockHash, requested.copy(received = true)) >>
-                  Log[F].info(
-                    s"Block ${PrettyPrinter.buildString(b.blockHash)} marked as received."
+  private def blockIsOk(
+      b: BlockMessage,
+      peer: PeerNode
+  ): F[Boolean] =
+    withCasper(
+      casper =>
+        for {
+          _ <- validatorId match {
+                case None => ().pure[F]
+                case Some(id) =>
+                  F.whenA(b.sender == ByteString.copyFrom(id.publicKey.bytes))(
+                    Log[F].warn(
+                      s"There is another node $peer proposing using the same private key as you. " +
+                        s"Or did you restart your node?"
+                    )
                   )
               }
-            }
-      } yield ()
-    } >> processUnfinishedParentsAndAddBlock(b)
-  }
+          lastFinalizedStateBlock = approvedBlock.candidate.block
+          // TODO might be more checks can/should be applied here
+          validShard   = lastFinalizedStateBlock.shardId.equalsIgnoreCase(b.shardId)
+          validFormat  <- Validate.formatOfFields(b)
+          validSig     <- Validate.blockSignature(b)
+          validVersion <- casper.getVersion.flatMap(Validate.version(b, _))
+          // TODO 50 is hardcoded value expirationThreshold from MultiParentCasperImpl. Move to conf
+          oldBlock = (ProtoUtil.blockNumber(b) <= ProtoUtil.blockNumber(
+            lastFinalizedStateBlock
+          ) - 50).pure[F]
```

Comment:
Expiration threshold should only be hardcoded on one place and use the same reference here.